src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -903,7 +903,15 @@ fn traits_implemented_by<'a>(
                 ty
             );
             // Fast path: if this is a primitive simple `==` will work
-            let saw_impl = impl_type == ty;
+            let saw_impl = impl_type == ty
+                || match (impl_type.kind(), ty.kind()) {
+                    (ty::Adt(def, _), ty::Adt(tydef, _)) => {
+                        debug!("adt def_id: {:?}", def.did);
+                        def.did == tydef.did
+                    }
+                    (ty::Foreign(def_id), ty::Foreign(tydef_id)) => def_id == tydef_id,
+                    _ => false,
```

Comment:
@camelid your intuition was right: this piece was necessary, but I was not sure at the time how to compare them, but I think this is the way: if they both having matching types, we compare the `def_id` alone, this allows users to write `Foo` without the generic type parameters, and still have it match with `Foo<T>`. That's my understanding.

Regarding the `Foreign`, I'm not sure what that means to be honest, any ideas how can I test that case too to avoid regressions again?

Reply:
I think that explanation sounds right.

The `Foreign` is an `extern type` I believe (feature gate `#![feature(extern_types)]`; issue #43467).

Reply:
@camelid I actually realised there are already tests for Foreign, and because the `ty::Foreign` only has one field, the `def_id`, doing a `==` covers what we are doing. The tests are passing without this case in the `match`. I removed them.

Reply:
https://github.com/rust-lang/rust/blob/master/src/test/rustdoc/intra-doc/extern-type.rs

Reply:
That test doesn't test trait impls on extern types, though. Can you please add a test case for that?

Reply:
> because the ty::Foreign only has one field, the def_id, doing a == covers what we are doing.

Hmm, what does having only one field have to do with it?

Reply:
@camelid wrote tests for impl traits of extern types.

regarding the other question, `Foreign` which is a `TyKind` has derived `Eq`, and as far as I understand about derived `Eq` of algebraic data types, the implementation is just applying `Eq` to all the fields of that type. In this case, `Foreign` is a tuple with a single field, the `DefId`, so when you do `foreign1 == foreign2`, it's identical to `def_id1 == def_id2`, that's all there is to compare.

Reply:
It's true that `TyKind` has derived `Eq`, but `Ty` manually implements it and uses pointer equality (because `Ty`s are interned).

What I was asking is why it's okay to skip manually equating `Foreign` but not `Adt`.

Reply:
@camelid ah I see. I'm not sure then... I suppose we have the same question about all the other types, not just Foreign

Reply:
Oh 🤦 

The reason it's necessary for `Adt`s is because we want to ignore generics since the `ty` we're matching against may just have placeholders (or similar). Can you add a comment about that?

Reply:
yes... that's what I mean with my initial comment on this piece https://github.com/rust-lang/rust/pull/92792#discussion_r782573159, I didn't get my message through well 😁

Reply:
Yeah, I think I read your message and understood partially but then got confused later 😆